fix #16185

cooldome opened the pull request:
fix #16185.

To simplify review:
The change made is the following piece of code:
```nim
 if dest.kind == nkSym and dest.sym == ri.sym:
    # rule (self-assignment-removal):
    result = newNodeI(nkEmpty, dest.info)
```
is replaced with 
```nim
if exprStructuralEquivalent(dest, ri, strictSymEquality = true):
  # rule (self-assignment-removal):
  result = newNodeI(nkEmpty, dest.info)
```
at the top of `moveOrCopy`. No other changes

cooldome committed:
fix #16185

Clyybber commented:
This is not correct, consider
```nim
import random
proc main =
  var s = @[@[1], @[2], @[3]]
  var i = 0
  s[rand(2)] = s[rand(2)]
  echo s

main()
```

cooldome committed:
fix test

cooldome commented:
@Clyybber,
Agree but what would be the correct fix:
```nim
s[rand(2)] = s[rand(2)]
```
can produce self-assingment as well.
Should we exclude only guaranted the same location and produce if statement for anything that MAY be a self-assignment.

cooldome committed:
fix comment

cooldome committed:
fix comment

cooldome commented:
Added a fix for @Clyybber comment

cooldome commented:
Added a fix for @Clyybber comment. I have reverted part.
```nim
if dest.kind == nkSym and dest.sym == ri.sym:
    # rule (self-assignment-removal):
    result = newNodeI(nkEmpty, dest.info)
```
because `isAnalysableFieldAccess` doesn't covert thread local and global

Clyybber commented:
>  Should we exclude only guaranted the same location and produce if statement for anything that MAY be a self-assignment.

`=copy` hook handles that for us. So only eliminate the self assignment if we can guarantee it.

cooldome committed:
better approach
